diff --git a/src/penguin/gen_config.py b/src/penguin/gen_config.py
index 04a72c0..be6fbde 100644
--- a/src/penguin/gen_config.py
+++ b/src/penguin/gen_config.py
@@ -248,9 +248,7 @@ def make_config(fs, out, artifacts, settings_path, timeout=None, auto_explore=Fa
     settings = {}
     user_settings = {}
     if settings_path:
-        # TODO: apply all use cases
-        relative_path = "src/penguin/resources/default_settings.yaml"
-        absolute_path = os.path.join(os.getcwd(), relative_path)
+        absolute_path = os.path.join(*[dirname(dirname(__file__)), "resources", "default_settings.yaml"])
         try:
             with open(settings_path, "r") as user_f:
                 user_settings = yaml.safe_load(user_f)
